File: app/src/main/java/com/sapienapps/christisus/planner/ManiClassPlanner.kt
package com.sapienapps.christisus.planner

import android.content.Context
import com.sapienapps.christisus.excelreader.FileUtilsV2

class ManiClassPlanner(
    override val maxClasses: Int,
    override val maxStudentsPerClass: Int,
    override val allowedProfileCombinations: List<List<Profile>>,
    override val allowedLanguageCombinations: List<List<Language>>
) : ClassPlanner {
    private val unassignedStudents = mutableListOf<Student>()
    private val classes = mutableListOf<ClassRoom>()

    override fun assignStudentsToClasses(students: List<Student>) {

        for (i in 0 until maxClasses) {
            val profileCombination = allowedProfileCombinations[i % allowedProfileCombinations.size]
            val languageCombination =
                allowedLanguageCombinations[i % allowedLanguageCombinations.size]
            classes.add(
                ClassRoom(
                    id = i + 1,
                    allowedProfiles = profileCombination,
                    allowedLanguages = languageCombination,
                    maxStudents = maxStudentsPerClass
                )
            )
        }

        val assignedStudents = mutableSetOf<Student>()
        val studentQueue = students.toMutableList()

        studentQueue.sortWith(compareBy({ it.profile }, { it.language }))

        while (studentQueue.isNotEmpty()) {
            val student = studentQueue.removeFirst()
            val possibleClasses = classes.filter { classRoom ->
                classRoom.students.none { it.nonFriendsList.contains(student.firstName) } &&
                        classRoom.allowedProfiles.contains(student.profile) &&
                        classRoom.allowedLanguages.contains(student.language)
            }

            val assignedClass = possibleClasses.minByOrNull { it.students.size }
            if (assignedClass != null) {
                if (assignedClass.students.size < maxStudentsPerClass) {
                    assignedClass.students.add(student)
                    assignedStudents.add(student)
                } else {
                    val removableStudent = assignedClass.students.find {
                        it.friendsList.isEmpty() || it.friendsList.none { friend -> assignedClass.students.any { s -> s.firstName == friend } }
                    }
                    if (removableStudent != null) {
                        assignedClass.students.remove(removableStudent)
                        assignedStudents.remove(removableStudent)
                        assignedClass.students.add(student)
                        assignedStudents.add(student)
                        studentQueue.add(removableStudent)
                    } else {
                        unassignedStudents.add(student)
                    }
                }
            } else {
                unassignedStudents.add(student)
            }
        }

        for (classRoom in classes) {
            for (student in classRoom.students.toList()) {
                val betterClass = classes.find {
                    it.students.size < maxStudentsPerClass &&
                            student.friendsList.any { friend -> it.students.any { s -> s.firstName == friend } }
                }
                if (betterClass != null && betterClass != classRoom) {
                    classRoom.students.remove(student)
                    betterClass.students.add(student)
                }
            }
        }
    }

    override fun optimizeClassAssignments() {
        var count = 0
        while (count < 10) {
            optimiseIteration()
            count++
        }

    }

    private fun optimiseIteration(){
        for (classRoom in classes) {
            val studentsWithFriends = classRoom.students.filter { it.friendsList.isNotEmpty() }

            for (student in studentsWithFriends) {
                val missingFriends = student.friendsList.filter { friend ->
                    classRoom.students.none { it.fullName() == friend || it.reverseFullName() == friend }
                }

                if (missingFriends.isNotEmpty()) {
                    val friendStudent = classes.flatMap { it.students }.find { student ->
                        student.fullName() in missingFriends || student.reverseFullName() in missingFriends
                    }
                    if (friendStudent != null) {
                        val friendClass = classes.find { it.students.contains(friendStudent) }
                        if (friendClass != null && friendClass != classRoom &&
                            classRoom.allowedProfiles.contains(friendStudent.profile) &&
                            classRoom.allowedLanguages.contains(friendStudent.language) &&
                            classRoom.students.none { it.nonFriendsList.contains(friendStudent.firstName) }
                        ) {
                            val swapCandidate = classRoom.students.find {
                                it.friendsList.isEmpty() &&
                                        it.nonFriendsList.isEmpty() &&
                                        friendClass.allowedProfiles.contains(it.profile) &&
                                        friendClass.allowedLanguages.contains(it.language)
                            }
                            if (swapCandidate != null) {
                                classRoom.students.remove(swapCandidate)
                                friendClass.students.add(swapCandidate)

                                friendClass.students.remove(friendStudent)
                                classRoom.students.add(friendStudent)
                            }
                        }
                    }
                }
            }
        }
    }
    override fun writeResultsToExcel(context: Context, outputPath: String,fileName:String): String {
        return FileUtilsV2.writeResultsToExcel(
            context,
            outputPath,
            classes = classes,
            conflictedStudents = unassignedStudents,
            fileName = fileName
        )
    }
}

File: christisus-web/src/planner/ManiClassPlanner.ts
import {
  ClassRoom,
  Language,
  Profile,
  Student,
  studentFullName,
  studentReverseFullName,
} from '../types/models';

const ITERATIONS_BUFFER = 80;
const CLASS_OPTIMISATION_ITERATIONS = 10;

export interface ClassPlannerResult {
  classes: ClassRoom[];
  unassignedStudents: Student[];
}

export class ManiClassPlanner {
  private readonly maxClasses: number;
  private readonly maxStudentsPerClass: number;
  private readonly allowedProfileCombinations: Profile[][];
  private readonly allowedLanguageCombinations: Language[][];
  private unassignedStudents: Student[] = [];
  private classes: ClassRoom[] = [];

  constructor(
    maxClasses: number,
    maxStudentsPerClass: number,
    allowedProfileCombinations: Profile[][],
    allowedLanguageCombinations: Language[][]
  ) {
    this.maxClasses = maxClasses;
    this.maxStudentsPerClass = maxStudentsPerClass;
    this.allowedProfileCombinations = allowedProfileCombinations;
    this.allowedLanguageCombinations = allowedLanguageCombinations;
  }

  assignStudentsToClasses(students: Student[]): void {
    for (let i = 0; i < this.maxClasses; i++) {
      const profileCombination =
        this.allowedProfileCombinations[i % this.allowedProfileCombinations.length];
      const languageCombination =
        this.allowedLanguageCombinations[i % this.allowedLanguageCombinations.length];
      this.classes.push({
        id: i + 1,
        students: [],
        allowedProfiles: profileCombination,
        allowedLanguages: languageCombination,
        maxStudents: this.maxStudentsPerClass,
      });
    }

    const assignedStudents = new Set<Student>();
    const studentQueue = [...students];

    studentQueue.sort((a, b) => {
      if (a.profile !== b.profile) {
        return a.profile.localeCompare(b.profile);
      }
      return a.language.localeCompare(b.language);
    });

    const maxIterations = ITERATIONS_BUFFER + students.length;
    let iteration = 0;

    while (studentQueue.length > 0 && iteration < maxIterations) {
      iteration++;
      const student = studentQueue.shift()!;

      const possibleClasses = this.classes.filter((classRoom) => {
        const hasNoNonFriend = !classRoom.students.some((s) =>
          s.nonFriendsList.includes(student.firstName)
        );
        const profileAllowed = classRoom.allowedProfiles.includes(student.profile);
        const languageAllowed = classRoom.allowedLanguages.includes(student.language);
        return hasNoNonFriend && profileAllowed && languageAllowed;
      });

      const assignedClass = possibleClasses.reduce<ClassRoom | null>((min, current) => {
        if (!min) return current;
        return current.students.length < min.students.length ? current : min;
      }, null);

      if (assignedClass) {
        if (assignedClass.students.length < this.maxStudentsPerClass) {
          assignedClass.students.push(student);
          assignedStudents.add(student);
        } else {
          const removableStudent = assignedClass.students.find(
            (s) =>
              s.friendsList.length === 0 ||
              !s.friendsList.some((friend) =>
                assignedClass.students.some((st) => st.firstName === friend)
              )
          );
          if (removableStudent) {
            const idx = assignedClass.students.indexOf(removableStudent);
            assignedClass.students.splice(idx, 1);
            assignedStudents.delete(removableStudent);
            assignedClass.students.push(student);
            assignedStudents.add(student);
            studentQueue.push(removableStudent);
          } else {
            this.unassignedStudents.push(student);
          }
        }
      } else {
        this.unassignedStudents.push(student);
      }
    }

    if (studentQueue.length > 0) {
      this.unassignedStudents.push(...studentQueue);
    }

    for (const classRoom of this.classes) {
      for (const student of [...classRoom.students]) {
        const betterClass = this.classes.find(
          (c) =>
            c.students.length < this.maxStudentsPerClass &&
            student.friendsList.some((friend) =>
              c.students.some((s) => s.firstName === friend)
            )
        );
        if (betterClass && betterClass !== classRoom) {
          const idx = classRoom.students.indexOf(student);
          classRoom.students.splice(idx, 1);
          betterClass.students.push(student);
        }
      }
    }
  }

  optimizeClassAssignments(): void {
    let count = 0;
    while (count < CLASS_OPTIMISATION_ITERATIONS) {
      this.optimiseIteration();
      count++;
    }
  }

  private optimiseIteration(): void {
    for (const classRoom of this.classes) {
      const studentsWithFriends = classRoom.students.filter(
        (s) => s.friendsList.length > 0
      );

      for (const student of studentsWithFriends) {
        const missingFriends = student.friendsList.filter(
          (friend) =>
            !classRoom.students.some(
              (s) =>
                studentFullName(s) === friend || studentReverseFullName(s) === friend
            )
        );

        if (missingFriends.length > 0) {
          const friendStudent = this.classes
            .flatMap((c) => c.students)
            .find(
              (s) =>
                missingFriends.includes(studentFullName(s)) ||
                missingFriends.includes(studentReverseFullName(s))
            );

          if (friendStudent) {
            const friendClass = this.classes.find((c) =>
              c.students.includes(friendStudent)
            );

            if (
              friendClass &&
              friendClass !== classRoom &&
              classRoom.allowedProfiles.includes(friendStudent.profile) &&
              classRoom.allowedLanguages.includes(friendStudent.language) &&
              !classRoom.students.some((s) =>
                s.nonFriendsList.includes(friendStudent.firstName)
              )
            ) {
              const swapCandidate = classRoom.students.find(
                (s) =>
                  s.friendsList.length === 0 &&
                  s.nonFriendsList.length === 0 &&
                  friendClass.allowedProfiles.includes(s.profile) &&
                  friendClass.allowedLanguages.includes(s.language)
              );

              if (swapCandidate) {
                const swapIdx = classRoom.students.indexOf(swapCandidate);
                classRoom.students.splice(swapIdx, 1);
                friendClass.students.push(swapCandidate);

                const friendIdx = friendClass.students.indexOf(friendStudent);
                friendClass.students.splice(friendIdx, 1);
                classRoom.students.push(friendStudent);
              }
            }
          }
        }
      }
    }
  }

  getResult(): ClassPlannerResult {
    return {
      classes: this.classes,
      unassignedStudents: this.unassignedStudents,
    };
  }
}
